### ISSUE TITLE
(RVI) Relative Volatility Index

### ISSUE THREAD

**Issue by mrjbq7** [TA-Lib/ta-lib#366]
> [!IMPORTANT]
> **RULED: ties go to neither bucket.** Group **Volatility Indicators**; **50.0** on a zero denominator. **Do not add the Tulip arm** — `beta/rvi.c` is not RVI and would bless wrong output.
>
> The `Open questions` section further down is **historical** — it records what was open
> when the card was written, not what is. The ruling comment on this issue is the current
> state and wins over the body wherever they disagree.

A request was made for the "Relative Volatility Index Indicator"

https://github.com/TA-Lib/ta-lib-python/issues/631

---

## Proposal

**SHIP-SCOPED.** Ship Dorsey's **1993 original** close-form as `TA_RVI` — two independent
periods, Wilder smoothing, ties counted as neither up nor down. Do **not** ship the 1995
"refined" high/low average in the same card.

There is zero residual doubt about *what the formula is*: an independent transcription of
trading-signals' arithmetic reproduces that library **bit for bit, at four parameter tuples**
(MEASURED, below), and a second library corroborates the smoother, the gate and the tie rule on
its converged tail. The shipped C body will still sit ~2.4e-13 from the golden — that is
deliberate house FP spelling, itself measured term by term. Two in-tree notes are therefore now
wrong and should be corrected when those files are retired:
`to_be_deleted/ta-coverage-gap-and-roadmap.md:167` files RVI under **"HARD — single or no
oracle"**, and `to_be_deleted/pandas-ta-classic-catalog.md:140` reduces it to "RSI applied to
stddev; up/down variant question".

**State this in `rvi.md` and in the issue reply:** the shipped function will match **neither**
implementation the requester named. `gh issue view 631` — the follow-up reads *"values of RVI
of chartiq and tradingview do not match... basically i want RVI for chartiq"*. ChartIQ's
built-in studies reference does not document a Relative Volatility study at all (**[F]**
`documentation.chartiq.com/tutorial-Using and Customizing Studies - Definitions.html`), so its
form is not determinable and cannot be an oracle. TradingView publishes no formula either
(**[F]**, below) and the one structural sentence it does publish describes the *high/low*
form. The variant decision is the whole job here, and the numbers are large: the published
things carrying this name sit **MEASURED up to 36.04 index points apart on a 0..100 scale.**

## What it is

RSI with the rolling **standard deviation** substituted for the price change: the sign of the
close-to-close change still routes the bar to the up or the down bucket, but the magnitude fed
in is the volatility, not the move. Donald Dorsey, *Technical Analysis of Stocks & Commodities*
**V.11:6 (Jun 1993) 253-256**; revised in **V.13:9 (Sep 1995) 388-391**, which averages the RVI
of the highs with the RVI of the lows.

**Two naming hazards, both real.**

1. **`RVI` vs `RVIorig`.** Two of the source families below reserve "RVI" for the **1995**
   high/low average and call the close-only 1993 form **`RVIorig`** — the Chart manual and
   DXcharts, both **[F]**. `TA_RVI` therefore names what those two call `RVIorig`. The choice
   is justified — pandas-ta-classic (`refined=False`) and trading-signals both default to
   close-only, and Dorsey's own note is that the closes are used when high/low are unavailable
   — but it must be stated in `rvi.md`, not left implicit.
2. **Relative *Vigor* Index.** MetaTrader 5's built-in "RVI" is an unrelated
   close-open/high-low oscillator. pandas-ta-classic and trading-signals both resolve the
   clash as this card does (`rvi`/`RVI` = volatility, `rvgi`/`RVGI` = vigor). No RVGI card
   exists on the board, so the abbreviation is free.

## Formula

`n1` = standard-deviation window (default 10), `n2` = Wilder smoothing period (default 14).

```
sigma_i = stddev of inReal[i-n1+1 .. i]
up_i    = (inReal[i] >  inReal[i-1]) ? sigma_i : 0
dn_i    = (inReal[i] <  inReal[i-1]) ? sigma_i : 0
U_i     = Wilder_n2(up_i), seeded with the simple average of the first n2 up_i
D_i     = Wilder_n2(dn_i), seeded the same way
RVI_i   = 100 * U_i / (U_i + D_i)
```

**Population vs sample stddev is not a decision and not a variant axis.** No published source
states which is meant (the Chart manual, **[F]**, does not), and it cannot matter: sample =
population x sqrt(n/(n-1)), a constant that cancels in `U/(U+D)` because both legs are linear
in sigma and the seeded Wilder recurrence is linear. **MEASURED** on the 252-bar corpus at
(14,10): switching ddof 0 -> 1 moves the result by **2.84e-14 abs / 5.93e-16 rel**, i.e. FP
noise. The same argument retires pandas' `scalar` and any `optInNbDev`-style multiplier. Use
`TA_STDDEV`'s existing population form and say nothing further about it.

TA-Lib-shaped. `VAR_STEP(i)` denotes `var.c`'s per-bar body (add incoming against `shift`,
mean, subtract trailing, reseed test); the point of copying it verbatim is that the composite
differential below is then bit-exact.

```c
lookbackTotal = (optInStdDevPeriod - 1) + (optInTimePeriod - 1)
              + TA_GetUnstablePeriod(TA_FUNC_UNST_RVI);
if( startIdx < lookbackTotal ) startIdx = lookbackTotal;
if( startIdx > endIdx ) { *outBegIdx = 0; *outNBElement = 0; return TA_SUCCESS; }

/* wAlpha is derived FROM wBeta, exactly as rma.c:51-52 does: that order is what
 * makes wAlpha + wBeta exactly 1, and it is the only spelling bit-equal to a
 * TA_RMA leg. The `1.0/n` spelling is algebraically equal and not bit-equal. */
wBeta  = (double)(optInTimePeriod - 1) / (double)optInTimePeriod;
wAlpha = 1.0 - wBeta;

/* TWO cursors, as var.c:61-73 has. `base` is the first bar the sigma window
 * touches; the first bar that HAS a sigma is n1-1 later. */
base         = startIdx - lookbackTotal;
trailingIdx  = base;
shift        = inReal[base];
periodTotal1 = periodTotal2 = 0.0;
for( j = base; j < base + optInStdDevPeriod - 1; j++ )   /* n1-1 terms */
   { tempReal = inReal[j] - shift; periodTotal1 += tempReal;
     tempReal *= tempReal;         periodTotal2 += tempReal; }

/* == startIdx - (optInTimePeriod-1) - unstable */
today = base + (optInStdDevPeriod - 1);

upTotal = dnTotal = 0.0;
for( i = 0; i < optInTimePeriod; i++, today++ ) {
    sigma = sqrt( VAR_STEP(today) );
    /* today-1 >= base always, because optInStdDevPeriod >= 2 is enforced by
     * the parameter range. That bound is load-bearing, not cosmetic: at n1 == 1
     * the minimum call reads inReal[-1]. */
    delta = inReal[today] - inReal[today-1];
    if     ( delta > 0.0 ) upTotal += sigma;
    else if( delta < 0.0 ) dnTotal += sigma;
}
prevUp = upTotal / optInTimePeriod;      /* seed bar is startIdx - unstable */
prevDn = dnTotal / optInTimePeriod;

/* Skip the unstable period: TA_GetUnstablePeriod(TA_FUNC_UNST_RVI) more steps
 * with no store (rma.c:74-81), each one
 *   STEP(today, &u, &d);
 *   prevUp = wAlpha*u + wBeta*prevUp;   prevDn = wAlpha*d + wBeta*prevDn;
 * ONE statement per leg -- splitting it unfuses the multiply-add and adds a
 * second latency to the recurrence's dependency chain (rma.c:54-58).
 * Then store at startIdx, and step-then-store for each later bar. */
outIdx = 0;
total  = prevUp + prevDn;
outReal[outIdx++] = (total == 0.0) ? 50.0 : 100.0 * (prevUp / total);
*outBegIdx = startIdx; *outNBElement = outIdx;
```

Lookback is `(n1-1) + (n2-1) + unstable`. **MEASURED** against trading-signals at four tuples,
exact at all four: `(n2,n1) = (14,10) -> beg 22`, `(30,30) -> 58`, `(1,10) -> 9`, `(2,2) -> 2`.

**`total == 0.0` is reachable on real data, not only on flat input.** At `optInTimePeriod = 1`
the recurrence has no memory, so any tie bar makes `U == D == 0` on its own. **MEASURED**: at
`(1,10)` on the shipped corpus, bar 101 (its single tie) emits exactly `50.0`. Return the
neutral centre — `imi.c:69` does exactly this, with exactly the `100.0*(U/(U+D))` spelling, and
`RelativeVolatilityIndex.js:36-37` reached the same constant independently. Division-first also
removes the only overflow path: `100.0*(U/T)` is bounded by 100 where `(100.0*U)/T` is not.

### Rejected variants — MEASURED on `TA_SREF_close_daily_ref_0_PRIV` (252 bars, `src/tools/ta_regtest/test_data.c:92`), max abs deviation from the proposed form

| Variant | Source | Max dev | Why rejected |
|---|---|---|---|
| tie (`close == prev`) -> **down** | Chart manual's `EMA[W14](U) / EMA[W14](S)` (an `S` denominator **is** ties-to-down: it forces `U+D == S`); DXcharts, same denominator; community Pine reproductions | **1.97** (at the tie, bar 101) | Not the literal RSI analogue. `rsi.c:140-146` routes a zero change to the `else` arm adding `0.0` — neither bucket. See Open question 1: this is a genuine split. |
| tie -> **up** | ProRealCode's ProRealTime port, `up=(src>=src[1])*selectedStdev` | **2.71** (bar 101) | Only source that does this. |
| plain **EMA** (`2/(n2+1)`) instead of Wilder | community Pine `ta.ema(...,14)`; pandas' `mamode` defaults to `"ema"` | **11.55** (ties->neither) / **11.84** (ties->down) | Four independent sources say Wilder: Chart manual `EMA[W14]` *"follows J. Welles Wilder's reckoning ... the same as an RSI"*; thinkorswim offers `Wilder's`; MotiveWave and eSignal both default `method = SMMA`. Dorsey's construction is RSI's. |
| **one** period for both legs (n1 = n2 = 14) + EMA | pandas-ta-classic default (`rvi.py:20,24` passes one `length` to both `stdev` and `ma`) | **15.59** | Every other source separates them (10 and 14). |
| 9-period stddev | fmlabs | **2.16** | One source against five (Chart manual, TradingView, DXcharts, MotiveWave, trading-signals) that all say 10. |
| **RSI of the stddev series** — route by `sign(sigma_i - sigma_{i-1})` and feed the *difference* | MotiveWave and eSignal both publish this pseudocode verbatim, under this name, with `smoothPeriod=14 / sdPeriod=10 / method=SMMA` | **35.15**, Pearson r **0.23** | A different indicator. It is `TA_RSI(TA_STDDEV(close,10),14)`, not Dorsey's. Recorded because those two pages are otherwise the card's best corroboration of the periods and the smoother — the same page can be right about the parameters and wrong about the formula. |
| linear-regression residual + hold-the-untaken-leg EMA | Tulip `beta/rvi.c` (see Oracles) | **36.04**, Pearson r **0.31** | Not RVI at all. |
| 1995 "refined" `0.5*(RVI(high)+RVI(low))` | Dorsey 1995; Chart manual; DXcharts; pandas `refined=True` | separate fn | Different input bundle. See Open question 4. |

## Proposed API

```c
TA_RetCode TA_RVI( int    startIdx,
                   int    endIdx,
                   const double inReal[],
                   int    optInTimePeriod,     /* From 1 to 100000, default 14 */
                   int    optInStdDevPeriod,   /* From 2 to 100000, default 10 */
                   int   *outBegIdx,
                   int   *outNBElement,
                   double outReal[] );

int TA_RVI_Lookback( int optInTimePeriod, int optInStdDevPeriod );
/* (optInStdDevPeriod-1) + (optInTimePeriod-1) + TA_GetUnstablePeriod(TA_FUNC_UNST_RVI) */
```

- Group **`Volatility Indicators`** (existing; joins ATR/NATR/TRANGE, `atr/atr.yaml:2`). Both
  libraries file it there — `rvi.py:106` sets `category = "volatility"`, trading-signals ships
  it at `dist/volatility/RVI/`. See Open question 3. Input bundle: plain `inReal`, single
  series (conventionally close). One output — no output-count blocker.
- **Parameter order is settled house convention, not a coin flip.** `kc/kc.yaml:10-15,17-22` is the
  exact structural precedent: `optInTimePeriod` = the primary smoothing period **first**
  (default 20, range `[2,100000]`), `optInATRPeriod` = the sub-indicator's own period
  **second** (default 10, range `[1,100000]`) — same shape, same asymmetric ranges, same
  `optIn<SubIndicator>Period` naming. thinkorswim corroborates that the two periods are
  independent by exposing both (`st dev length`, `average length`).
- `optInTimePeriod` = the **Wilder** period, default **14**. Sources: Chart manual
  (`EMA[W14]`); DXcharts (`14-days EMA`); MotiveWave (`Smoothing Period 14`); eSignal
  (`smoothPeriod default 14`); trading-signals `interval = 14`
  (`RelativeVolatilityIndex.js:10`). Range `[1, 100000]`, matching `rma/rma.yaml:13` — the
  smoother *is* RMA, `n2 == 1` is well defined (`wAlpha, wBeta` = exactly `(1, 0)`), and it is
  the tie-rule probe below. No `period1_identity` flag: at `n2 == 1` the output is not the
  input.
- `optInStdDevPeriod`, default **10**. Sources: Chart manual (`S = Stddev[10 days]`); DXcharts
  (10-day); MotiveWave (`Standard Deviation Period 10`); eSignal (`sdPeriod default 10`);
  TradingView (`Length ... 10 by default`); trading-signals `stddevInterval = 10`. Range
  `[2, 100000]` — the bound `stddev/stddev.yaml:13` uses, trading-signals throws below 2
  (`:12-14`), and `n1 >= 2` is what keeps `inReal[today-1]` in bounds at the minimum call.
- Flags `[unstable_period, stream]` — the pair `rsi/rsi.yaml:4` and `atr/atr.yaml:4` carry. Not
  `overlap` (0..100, not price-scaled), not `nan_inf_output`, and **not `path_dependent`**
  (#127): that flag is carried by AD/ADOSC/NVI/OBV/PVI/SAR/SAREXT/SUPERTREND/VWAP/WAD, whose
  state accumulates from bar 0. RVI's seed is anchored at `startIdx - lookbackTotal`, exactly
  like RSI.
- New `TA_FUNC_UNST_RVI`. **Append at 25** after `RMA = 24` (`input/enums.yaml:76`), per that
  file's own APPEND rule. Four reserved slots are marked reusable (`UNUSED_1/12/15/22`,
  `enums.yaml:53,64,67,74`); reusing one is also not an insertion, but they are the ABI
  gravestones of *retired* knobs and re-pointing one at a live function makes
  `TA_FuncUnstId` unreadable. Appending costs **seven hand edits beyond the enum**, one of
  them a C# `MetadataTest` literal count that neither `regen-check` nor `cargo test` can see —
  it fails ~15 minutes into `build.py servers`.

## References

**[F]** = fetched this session, HTTP 200. **[C]** = cited, not fetched.

- **[C]** Dorsey, "The Relative Volatility Index", *TASC* V.11:6 (Jun 1993) 253-256,
  https://store.traders.com/-v11-c06-therela-pdf.html ; "Refining the Relative Volatility
  Index", V.13:9 (Sep 1995) 388-391, https://store.traders.com/-v13-c09-refinin-pdf.html
  (both paywalled; titles/pagination from the store listings).
- **[F]** Chart manual (Kevin Ryde),
  https://user42.tuxfamily.org/chart/manual/Relative-Volatility-Index.html — verbatim:
  `S = Stddev[10 days]`; `U = S if price > prev price, 0 otherwise`;
  `RVIorig = 100 * (EMA[W14] of U) / (EMA[W14] of S)`;
  `RVI = (RVIorig of highs + RVIorig of lows) / 2`; *"The EMA period follows J. Welles Wilder's
  reckoning ... the same as an RSI"*; *"When a data source doesn't provide high/low values just
  the closes are used, which ends up as Dorsey's original."* It does **not** specify population
  vs sample. **The single most load-bearing reference on this card.**
- **[F]** DXcharts / Devexperts,
  https://devexperts.com/dxcharts/kb/docs/relative-volatility-index-rvi — same `RVI`/`RVIorig`
  split, `RVIorig = 100 * (14-days EMA of U) / (14-days EMA of SD)`, 10-day stddev. The `SD`
  denominator implies ties -> down.
- **[F]** TradingView,
  https://www.tradingview.com/support/solutions/43000594684-relative-volatility-index/ —
  resolves; publishes **no formula, no MA type, no tie rule**. Confirms `Length ... 10 by
  default`, and says the RVI *"uses **high and low** price standard deviation instead of the
  RSI's method of absolute change in price"* — i.e. its own prose points at the 1995 form. It
  is **not** a witness for either the plain-EMA or the ties-to-down choice; those come from
  community Pine reproductions only.
- **[F]** thinkorswim / Schwab,
  https://toslc.thinkorswim.com/center/reference/Tech-Indicators/studies-library/R-S/RelativeVolatilityIndex
  — parameters `st dev length`, `average length`, `average type` in {simple, exponential,
  weighted, **Wilder's**, Hull}. No defaults published. The strongest external evidence that
  the two periods are independent and the smoother is a genuinely open choice.
- **[F]** MotiveWave, https://docs.motivewave.com/studies/q-r — `Smoothing Period 14`,
  `Standard Deviation Period 10`, `Method SMMA`, input `close`; but its published pseudocode
  is the RSI-of-stddev variant rejected above.
- **[C]** eSignal EFS,
  https://efs.kb.esignal.com/hc/en-us/articles/6362864396059-Relative-Volatility-Index-RVI —
  403 to a plain fetch; its pseudocode is reproduced verbatim inside the MotiveWave page and is
  the same RSI-of-stddev form, `default is SMMA / smoothPeriod 14 / sdPeriod 10 / input close`.
- **[F]** fmlabs, https://www.fmlabs.com/reference/RVI.htm — says **9**-period stddev. Outlier;
  recorded so the 10 is a decision, not an oversight.
- **[F]** ProRealCode,
  https://www.prorealcode.com/prorealtime-indicators/relative-volatility-index-rvi/ — quoted in
  the variants table. Its smoother is `average[smoothLength,1]`; the PRT type code was not
  verified, so it is cited for the periods and the tie rule only.
- **[F]** ChartIQ built-in studies reference,
  https://documentation.chartiq.com/tutorial-Using%20and%20Customizing%20Studies%20-%20Definitions.html
  — **no Relative Volatility entry**. The requester's target form is not publicly determinable.
- **[C]** MetaTrader 5 help, *Relative Vigor Index*,
  https://www.metatrader5.com/en/terminal/help/indicators/oscillators/rvi
- Read directly, this session: `pandas_ta_classic/volatility/rvi.py:18-32,60-108`,
  `overlap/rma.py:26`, `statistics/stdev.py:21`, `utils/_core.py:173-177` (0.6.52);
  trading-signals 8.3.0 `dist/volatility/RVI/RelativeVolatilityIndex.js:10-42`,
  `dist/trend/WSMA/WSMA.js:11,20-23`, `dist/trend/SMA/SMA.js`,
  `dist/util/math/getStandardDeviation.js:2-7`, `getAverage.js:2`; Tulip 0.9.2
  `beta/rvi.c:26-100`, `build.tcl:112`.

## Sample input/output (board policy)

Input: `TA_SREF_close_daily_ref_0_PRIV`, 252 bars (`src/tools/ta_regtest/test_data.c:92`;
`ta-lib-oracles/trading_signals_serve/corpus.json` is byte-equal to it). Parameters
`optInTimePeriod = 14`, `optInStdDevPeriod = 10`. **MEASURED**: `outBegIdx = 22`,
`outNBElement = 230`.

Left column: the proposed C body's spelling (`var.c` shifted-sum + reseed, `rma.c` `wBeta`-first
coefficients, `imi.c` `100.0*(U/T)` combine), computed in Python at double precision. Right
column: trading-signals 8.3.0, which is the value to freeze.

| bar | TA_RVI (proposed spelling) | trading-signals 8.3.0 | trading-signals hex |
|---|---|---|---|
| 22 | 52.713741724623162 | 52.713741724623155 | `404a5b5be38a8607` |
| 23 | 48.646868138730177 | 48.646868138730177 | `404852cc933e55d3` |
| 24 | 52.943714669251996 | 52.943714669251996 | `404a78cba46c98ad` |
| 25 | 48.411818272631038 | 48.411818272631038 | `404834b6760e6c3a` |
| ... | | | |
| 249 | 53.269833208990192 | 53.269833208990150 | `404aa289e503fe64` |
| 250 | 52.432431467108465 | 52.432431467108586 | `404a3759ea107f10` |
| 251 | 51.398767922312992 | 51.398767922313233 | `4049b30ad3c883bd` |

**MEASURED over all 230 values: max abs 2.42e-13, max relative 4.87e-15.** The whole residual
is three deliberate house FP-spelling choices, each isolated (all else equal, at (14,10)):

| spelling difference | max abs | max rel | bit-equal |
|---|---|---|---|
| `var.c` shifted-sum + reseed vs the two-pass stddev | 2.56e-13 | 4.98e-15 | 40/230 |
| `wBeta`-first vs WSMA's `(x-prev)*(1/n) + prev` | 2.84e-14 | 6.05e-16 | 81/230 |
| `100.0*(U/T)` vs `(100.0*U)/T` | 1.42e-14 | 2.83e-16 | 127/230 |

The stddev algorithm dominates; it sets the tolerance.

## Oracles

| Oracle | Covers the proposed form? | Arm status | Work to add the arm |
|---|---|---|---|
| **trading-signals 8.3.0** | **Yes, exactly** — `volatility/RVI`: Wilder via `WSMA`, separate `stddevInterval`/`interval`, ties to neither, 50 on a zero total | **must be added** | one `ARMS` row in `trading_signals_serve/capture.mjs` (`ctor: p => new ts.RelativeVolatilityIndex({interval:p[0], stddevInterval:p[1]})`) plus a `close`-only `feed` case — `bar()` at `capture.mjs:59-61` builds only `hl`/`ohlcv`. ~6 lines. Arm **#4**, which trips that README:14's own *"promote it to a server if a third or fourth arm lands here"*. |
| **pandas-ta-classic 0.6.52** | **On the converged tail, yes** — `mamode='rma'`, `n1 == n2 == length`, ties to neither | **must be added** | one `compute_RVI` + one `SPECS` row in `pandas_serve/server.py`, the `compute_CMOU` shape (`server.py:405-413,523`), ~10 lines. Compare only `length in {3,5}` and only bars `>= 2(length-1) + 168`. |
| **Tulip Indicators 0.9.2** | **No — and it is buildable, so say so loudly** | **DO NOT ADD** | see below |
| **ta4j** 0.22.6 and 0.24.1 | **No** | impossible | Verified by listing both jars under `org/ta4j/core/indicators/`: no Relative Volatility and no Relative Vigor class. `volume/RelativeVolumeStandardDeviationIndicator` is relative *volume*. |

**Tulip — the card's live instance of the silently-wrong-oracle hazard.**
`tulip_serve/vendor/tulipindicators/beta/rvi.c` is a complete 252-line implementation
(`ti_rvi_start`, `ti_rvi`, `ti_rvi_stream_new/_free/_run`), and adding it is **three lines**:
`build.sh:48-69` already carries an explicit beta allow-list (`BETA_SRCS=(beta/cmf.c
beta/smi.c)`), and `beta/rvi.c` is exactly the `smi.c` case that comment describes — it names
`TI_INDICATOR_RVI_INDEX` only inside `ti_rvi_stream_new`, so `-DTI_INDICATOR_RVI_INDEX=0`
builds it. **MEASURED** (compiled and run this session, `ti_rvi(sma_period=14,
stddev_period=10)`, same 252 bars): **max abs 36.04 index points, mean |d| 12.37, Pearson r
0.31**, `beg=9 nb=243` against our `beg=22 nb=230`. It is not RVI:

- `beta/rvi.c:61-64` fits `y = a + bx` over the window and takes the last point's residual,
  `higher = real[i] - (a + b*stddev_period)` — a linear-regression residual, then squared and
  divided by the period. Not a standard deviation.
- `beta/rvi.c:88-92` updates **only the taken leg's** EMA per bar; the untaken leg *holds*
  instead of decaying toward 0. Dorsey and both real oracles feed 0 into the untaken leg every
  bar.
- The EMAs are unseeded, and `ti_rvi_start` returns `stddev_period-1` with no smoothing
  lookback at all.
- Upstream tags the row `;#TODO suspect` (`build.tcl:112`) and never builds it.

`composite-functions.md` §9.8 names the failure mode: *"a silently-wrong oracle blesses wrong
output"*. Adding this arm would do exactly that. Its only value is corroborating the two-period
signature (`build.tcl:112` plans `{period {smooth_period}}`).

**pandas — why it is a real second arm, not a negative control.** `overlap/rma.py:26` seeds
from `close.iloc[0:length].mean()` over a NaN-prefixed input, so it emits from bar `length-1`
off a **one-sample** seed (MEASURED at `length=14`: `beg=13`, first values `0.0, 7.2487,
13.9552`). But the Wilder transient decays as `((L-1)/L)^k`, so at small `L` it converges onto
a correctly SMA-seeded Wilder of the same `(n1 == n2, ties -> neither)` shape. **MEASURED**, max
abs deviation from the proposed form over the bars from the given offset onward:

| length | skip 0 | skip 56 | skip 112 | skip 168 |
|---|---|---|---|---|
| 3 | 6.79 | 2.05e-9 | 8.95e-12 | **5.19e-12** |
| 5 | 10.44 | 3.68e-5 | 1.18e-10 | **1.54e-12** |
| 10 | 4.94 | 5.83e-3 | 2.26e-5 | 2.07e-8 |
| 14 | 11.91 | 1.10e-1 | 2.13e-3 | 3.78e-5 |

That arm independently witnesses the **Wilder smoother**, the **up/down gate** and the **tie
rule** from a different codebase and a different language. It does *not* witness `n1 != n2`
(`rvi.py:20,24` passes one `length` to both legs) — only trading-signals does. Note also that
pandas' own docstring (`rvi.py:127`) says `DOWN = STDEV IF src.diff() <= 0` while its code
(`utils/_core.py:173-177`) routes ties to neither: **its documentation contradicts its own
implementation on the tie.**

## Verification plan

1. **Composite differential**, bit-exact, appended to
   `src/tools/ta_regtest/ta_test_func/test_composite2.c` (705 lines; the natural target — there
   is no `test_composite.c`, and appending avoids registering a new file in both
   `CMakeLists.txt` and `Makefile.am`). Reference: `TA_STDDEV(inReal, n1, 1.0)` into scratch;
   gate on `sign(inReal[i] - inReal[i-1])` into `up[]`/`dn[]`; `TA_RMA(up, n2)`,
   `TA_RMA(dn, n2)`; combine. Four gate details, each of which silently breaks bit-equality:
   - **`TA_STDDEV`'s `startIdx` must be `TA_RVI`'s own `startIdx - (n2-1) - unstable`**, not 0.
     `var.c:61-62,79` anchors both `shift = inReal[trailingIdx]` and the `barsSinceReseed`
     counter at the *call's own* start; a differently-anchored call reaches the same window by
     a different add/subtract sequence and separates at ~1e-13. With that startIdx it is
     bit-exact at *any* `startIdx`, not only the full-range call.
   - `TA_SetUnstablePeriod(TA_FUNC_UNST_RMA, TA_GetUnstablePeriod(TA_FUNC_UNST_RVI))` on the
     reference, or the legs seed at different bars and nothing lines up.
   - The fused body must spell the coefficients `wBeta` first, `wAlpha = 1.0 - wBeta`
     (`rma.c:51-52`).
   - The **combine** must be copied too: `100.0*(U/T)` vs `(100.0*U)/T` are MEASURED 2.83e-16
     apart, 127/230 bit-equal.

   Grid `(n1,n2)` over `{2,3,10,14,30} x {1,2,14,30}`.
   *Stated as a reading:* `composite-functions.md` §2 requires every decomposition **piece** to
   be a shipped function, and the up/down gate is not one. A compare and a select introduce no
   numerical logic, so the reference stays free of novel arithmetic — but that is an
   interpretation of §2, not what §2 literally says, and it should be signed off with the card.
2. **External golden — the only formula check.** Freeze the trading-signals values as
   hex-of-IEEE-bits at **four** tuples, `(n2,n1) = (14,10)`, `(30,30)`, `(1,10)`, `(2,2)`, and
   compare with `checkOracleValue(got, want, /*relTol*/ 1e-10, /*absTol*/ 1e-12)`
   (`ta_test_priv.h:294-319`). **Both terms are required and the numbers set them:**
   - relTol is driven by `(2,2)`, not `(14,10)`. **MEASURED** max relative deviation: 4.87e-15
     at (14,10), 1.18e-15 at (30,30), **1.82e-12 at (2,2)** — a 2-bar window is the worst case
     for `var.c`'s shifted-sum spelling. A 1e-12 relative tolerance would fail there. 1e-10
     leaves 55x.
   - absTol is driven by `(1,10)`, where `want` is exactly `0.0` on many bars and relative
     error is undefined. **MEASURED** max abs deviation there: **1.42e-14** (trading-signals'
     `WSMA` spells its step `(x - prev)*(1/n) + prev`, which is not exact at `n == 1`, where
     `rma.c`'s `wAlpha*x + wBeta*prev` with `(1.0, 0.0)` is). 1e-12 leaves 70x.

   Per `composite-functions.md` §4 the differential proves the optimization, not the formula;
   only this catches a wrong `n1`, a wrong tie rule, or an EMA where Wilder belongs.
3. **Second oracle — pandas converged tail.** `ta.rvi(close, length=L, mamode='rma')` for
   `L in {3,5}`, compare only bars `>= 2(L-1) + 168`, `checkOracleValue(relTol 1e-9, absTol
   1e-9)` (MEASURED worst 5.19e-12 and 1.54e-12 — ~200x and ~650x headroom). Freeze the
   crossover behaviour too: record that at `skip 0` the same arm is 6.79 and 10.44 away, so a
   future reader cannot mistake the offset for arbitrary.
4. **Frozen negative controls,** recorded in `rvi.md` so the next person does not "fix" them:
   pandas' default `ta.rvi(close, 14)` deviates from `TA_RVI(close, 14, 10)` by up to **15.59**;
   Tulip's `beta/rvi.c` by up to **36.04**; MotiveWave's/eSignal's published pseudocode by up to
   **35.15**. These are variant, not error.
5. **Free coverage** once the YAML lands: `--codegen` across C/Rust/Java/.NET,
   `--xlang-hash`/`server_verify` at zero tolerance, `--fuzz-064`, the #94 boundary sweep under
   ASan/UBSan, output-aliasing rejection (#108), and the unstable-period range sweep. That last
   one needs an **explicit stability class**, which `doRangeTestEx` requires and cross-checks
   against the unstable id (`ta_test_priv.h:235-252`): RVI is **`TA_STABLE_CONVERGING`**
   (RSI/ATR/EMA precedent — an IIR whose residual the unstable period bounds). One id, so
   `doRangeTestEx`, not `doRangeTestMulti`.
6. **The tie rule has an exact, non-decaying probe already in the corpus — use it instead of a
   hand vector.** At `optInTimePeriod = 1` the recurrence has no memory, so the single tie bar
   reads a literal. **MEASURED** at `(1,10)`, bar 101: ties->neither gives exactly **50.0**,
   ties->down exactly **0.0**, ties->up exactly **100.0**, and the whole 243-value series takes
   only the three values `{0.0, 50.0, 100.0}`. Assert those three literals and that bar.
   At `optInTimePeriod = 14` the same flip is visible but **decays**, and stating the curve
   matters because "the recurrence never forgets it" invites a tail spot-check that is nearly
   blind: **MEASURED** ties->down minus proposed = 1.9734 (bar 101), 1.7792 (102), 0.7064
   (120), 0.0766 (150), 1.72e-3 (200), **5.33e-5 (bar 251)**, decaying as `(13/14)^k`. A golden
   over all 230 values catches it (5.33e-5 >> tolerance); the last few bars alone barely do.
7. **Hand-written edge cases:**
   - **Flat input**: `sigma == 0` on every bar, so `U + D == 0` on every bar. Assert exactly
     `50.0` and `TA_SUCCESS` with no NaN (#112).
   - **`n2 == 1`**: the `{0.0, 50.0, 100.0}` assertion above, which doubles as the #112 guard's
     second firing path.
   - **In-place `inReal == outReal`**, and it is *tight*: at `n2 == 1` with unstable 0,
     `startIdx == n1-1`, so on iteration `k` the sigma window is exactly `[k .. n1-1+k]` while
     `outReal[0..k-1]` has already overwritten `inReal[0..k-1]`. Zero slack. `var.c`'s reseed
     rescan (`var.c:118`, `windowStart = i - nbInitialElementNeeded`) reads no lower, and the gate's `inReal[today-1]` sits at
     `n1-2+k >= k` for `n1 >= 2`. The invariant is that every read of a bar precedes the write
     landing on it; ASan cannot see a violation. Dedicated test.
   - `startIdx == endIdx == lookbackTotal`; `nbBars < lookbackTotal` (`TA_SUCCESS`,
     `outNBElement == 0`, `outBegIdx == 0`); `n1 == 2`, the only `n1` where the gate's
     `inReal[today-1]` sits at the sigma window's own left edge.
8. **No NaN/Inf can escape a successful call.** The zero-denominator is guarded above. The only
   other route would be overflow into `Inf/Inf`, and it is unreachable at the #94 sweep's
   `[TA_REAL_MIN, TA_REAL_MAX] = +/-3e37` (`include/ta_defs.h:115-116`): the widest window
   spread is 6e37, so `periodTotal2 <= n1*(6e37)^2 <= 3.6e80` at `n1 = 100000`, `sigma <=
   6e37`, `U + D <= 1.2e38`, and the division-first combine caps the result at 100. Every
   intermediate stays finite.

## Open questions

1. **Tie rule.** Recommend **neither bucket**: both real implementations do it (`RelativeVolatilityIndex.js:28-31`,
   pandas `utils/_core.py:173-177`), and it is the literal RSI analogue (`rsi.c:140-146`
   adds `0.0` for a zero change). Against it, **three** published descriptions imply ties -> down
   (Chart manual and DXcharts, both via an `EMA(SD)` denominator that forces `U + D == S`; and
   community Pine's `ta.change(src) <= 0`). So the honest tally is **2 implementations
   (neither) vs 3 descriptions (down)**, not a 2-2 split. Worth 1.97 index points at the tie.
   **A human should sign this off** — it is the one place where what is running and what is
   written disagree.
2. **Group.** `Volatility Indicators` (both libraries' own category, and the name) vs.
   `Momentum Indicators` (its actual shape — a bounded 0..100 oscillator next to RSI/CMO;
   the existing Volatility group is `{ATR, NATR, TRANGE}`, all unbounded range measures).
   Recommend Volatility; it is a user-facing grouping, and both external libraries agree.
3. **Parameter name only** — `optInStdDevPeriod` vs `optInStdDevTimePeriod`. The *order* is
   settled by `kc/kc.yaml:10-15,17-22`.
4. **The 1995 refined form.** Dorsey's own revision is `0.5*(RVI(high) + RVI(low))`, and two of
   the reference families reserve the bare name "RVI" for it. Deferred, not rejected: it needs
   an HL bundle, and pandas is the only library shipping it (`refined=True`, `rvi.py:37-44`) —
   with the wrong smoother and a single length, so not an oracle. A trading-signals arm *could*
   serve it by running two instances and averaging, the same "neutralise the harness, not the
   library" move the VWAP arm already makes. Second card (`RVIR`), not a flag here.
5. **Zero-denominator return.** `50.0` (recommended: `imi.c:69` after #112, and
   trading-signals independently) vs `0.0` (`rsi.c:170`, legacy). It is a shipped constant
   either way, and it is reachable on the real corpus (bar 101 at `optInTimePeriod = 1`), not
   only on synthetic flat input.

## Notes

- **Streaming: `[stream]` is expected but has no precedent for the *combination*, and that is
  the one thing to confirm before promising it.** Each construct streams alone — `var.c`'s
  in-loop reseed rescan (`stddev/stddev.yaml:4`, `var/var.yaml:4`, `bbands/bbands.yaml:4` all
  carry `[stream]`) and a Wilder recurrence (`atr`, `natr`, `rma`, `supertrend`) — but the two
  sets are **disjoint**: `grep -l 'wBeta\|prevRMA' input/*/*.c` gives atr, rma, natr,
  supertrend; `grep -l barsSinceReseed` gives var.c alone. No shipped body has both. Run
  `cargo run -- stream-census` on the new IR before the YAML claims `[stream]`. Ring is `n1`
  closes; the sigma window already holds `inReal[i-1]`, so the up/down gate costs no extra slot.
- **The unstable period is mandatory, not optional.** The Wilder seed error decays by
  `(n2-1)/n2` per bar and never reaches zero — MEASURED as the pandas mis-seed transient above:
  11.91 -> 1.10e-1 (56 bars) -> 3.78e-5 (168) at `length=14`.
- **ABI:** one new function; there is no C `TA_FuncID` — the ordinal ABI is the Rust crate's `FuncId` (`ta_codegen/output/rust/library/src/abstract_api.rs:44`), where the new variant renumbers every later discriminant, new `TA_FUNC_UNST_RVI = 25`. `RVI` sorts between `RSI` and
  `SAR` in `ta_codegen/input/` (currently `rsi`, `sar`); confirm `ta_ref_serve` still builds — a
  new name sorting into the frozen table broke it once (#228, AC).
- Gotchas that apply here: run a **full** `cargo run -- generate` (a `--func=RVI` filter skips
  `Core.java` and the servers); `scripts/build.py servers`, not `build`; add the CHANGELOG
  entry; and `ta_regtest --function=RVI` matches the *group tag*, so confirm the filter ran
  tests instead of exiting 0 vacuously.

**Comment by mario4tier** [comment]
## Ruling: ties go to neither bucket. Moved to Todo.

A tied bar (`close[i] == close[i-1]`) feeds **neither** the up nor the down accumulator. This follows
both real implementations (`RelativeVolatilityIndex.js:28-31`, pandas `utils/_core.py:173-177`) and is
the literal RSI analogue — `rsi.c:140-146` adds `0.0` for a zero change.

**TA-Lib/ta-lib-proposal-drafts#68 moves the tally and does not change this.** That card found `App::Chart`'s `RVI.pm:118-120` as a
third implementation, ties→down, making it 2 implementations (neither) vs 1 implementation + 2
descriptions (down) rather than the 2-vs-3 recorded here. The ruling above stands with that evidence
in hand. `rvi.md` states the divergence — it is worth 1.97 index points at the tie here and 4.64 in
TA-Lib/ta-lib-proposal-drafts#68's refined form.

The card's other two recommendations are accepted as written and were not separately disputed:
group **Volatility Indicators** (both libraries' own category, and the name), and **50.0** on a
zero denominator (`imi.c:69` after #112, and trading-signals independently) rather than `rsi.c:170`'s
legacy `0.0`. Both are reachable on the real corpus, so both are gated, not theoretical.

Do **not** add the Tulip arm. `beta/rvi.c` is buildable and is not RVI — it is a linear-regression
residual with a hold-instead-of-decay bug and unseeded EMAs; the card measured max abs 36.04 index
points against the proposed form. Adding it would be a silently-wrong oracle blessing wrong output.

**Comment by mario4tier** [comment]
### Stale in the Oracles section: the close-only feed already exists

The trading-signals row says the arm needs *"a close-only feed case — `bar()` at `capture.mjs:59-61`
builds only `hl`/`ohlcv`. ~6 lines."*

**That is already done.** `capture.mjs` now has a three-way feed and two shipped arms use the close
one:

```js
const bar = (i) => arm.feed === 'close'
   ? close[i]
   : arm.feed === 'hl'
   ? { high: high[i], low: low[i] }
   : { open: open[i], high: high[i], low: low[i], close: close[i], volume: volume[i] };
```

`VHF` and `FOSC` both declare `feed: 'close'`. The harness is also at **5 arms** (AC, AO, VWAP, VHF,
FOSC), not the 3 the card records — so the README's *"promote it to a server if a third or fourth arm
lands here"* threshold has already been passed and is a live question for whoever adds the sixth,
independent of this card.

So the RVI arm is one `ARMS` entry with `feed: 'close'`. The `~6 lines` of harness widening are not
owed.

**Comment by mario4tier** [comment]
> [!IMPORTANT]
> **RULED: RVI may be the first body combining the two streaming families.**

The concern was that RVI would be the first function whose body carries both:

- the **reseed-rescan** family (`barsSinceReseed` — `var`, `bbands`, `wma`, `hma`, `correl`, `beta`, `linearreg`/`_slope`/`_angle`/`_intercept`, `tsf`), and
- the **Wilder** family (`wBeta` / `prevRMA` — `atr`, `rma`, `natr`, `supertrend`).

Those two sets are currently **disjoint** — verified by grep over `ta_codegen/input/`. Each construct streams cleanly on its own and the added state is additive, but the combination has no precedent, so it sets a direction for the streaming analyzer rather than just for this indicator. That is why it was held for a human call rather than shipped on a card recommendation.

Approved. RVI ships with both, and becomes the precedent.

Implementation notes that stay binding:
- `optInStdDevPeriod >= 2` is load-bearing, not cosmetic: the up/down gate reads `inReal[today-1]`, and at the minimum call `today-1` lands exactly on `base` only because the sigma window is at least two bars wide. At `n1 == 1` the minimum call reads `inReal[-1]`. Range stays `[2, 100000]`.
- `var.c`'s per-bar step is transcribed **verbatim** — the shifted sums, `invPeriod`, `barsSinceReseed = 32*optInTimePeriod`, **both** floors (the 1e-6 reseed trigger and the 1e-12 post-reseed clamp, whose constants deliberately differ), and the "re-remove the trailing value under the new shift" tail. A plain two-pass substitution is not equivalent.

**Comment by mario4tier** [comment]
Shipped on `dev`: 8c0fedbcd

The first body combining the reseed-rescan family with a Wilder-smoothed pair, as ruled here. Its steady loop is VAR's paced form, not RSI's countdown: the stream analyzer removes the loop counter and demands exactly one surviving input cursor, and a countdown leaves four. The reshape moves no value — an 8120-line hex-of-IEEE-bits dump is byte-identical before, after, and after the full regenerate; Open/Update vs batch is 27406 compares with 0 mismatches.
